sentry-ruby/lib/sentry/profiler.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+
+return unless defined?(StackProf)
+
+require 'securerandom'
+
+module Sentry
+  class Profiler
+
+    VERSION = '1'
+    PLATFORM = 'ruby'
+    # 101 Hz in microseconds
+    DEFAULT_INTERVAL = 1e6 / 101
+
+    attr_reader :sampled, :started, :event_id
+
+    def initialize
+      @event_id = SecureRandom.uuid.delete('-')
+      @started = false
+      @sampled = nil
+    end
+
+    def start
+      return unless @sampled
+
+      @started = StackProf.start(interval: DEFAULT_INTERVAL,
+                                 mode: :wall,
+                                 raw: true,
+                                 aggregate: false)
+
+      @started ? log('Started') : log('Not started since running elsewhere')
+    end
+
+    def stop
+      return unless @sampled
+      return unless @started
+
+      StackProf.stop
+      log('Stopped')
+    end
+
+    def profile_context
+      { profile_id: @event_id }
+    end
+
+    # Sets initial sampling decision of the profile.
+    # @return [void]
+    def set_initial_sample_decision(transaction_sampled)
+      unless Sentry.initialized? && Sentry.configuration.profiling_enabled?
+        @sampled = false
+        return
+      end
+
+      unless transaction_sampled
+        @sampled = false
+        log('Discarding profile because transaction not sampled')
+        return
+      end
+
+      profiles_sample_rate = Sentry.configuration.profiles_sample_rate
+
+      case profiles_sample_rate
+      when 0.0
+        @sampled = false
+        log('Discarding profile because sample_rate is 0')
+        return
+      when 1.0
+        @sampled = true
+        return
+      else
+        @sampled = Random.rand < profiles_sample_rate
+      end
+
+      log('Discarding profile due to sampling decision') unless @sampled
+    end
+
+    def to_hash
+      return nil unless Sentry.initialized?
```

Comment:
In other parts of the SDK, `to_hash` always return `Hash`. Can we return empty hash here instead? I'd be surprising to see `to_hash` return `nil` IMO.